### PR TITLE
Copter: Add support for MAV_CMD_DO_PAUSE_CONTINUE in GUIDED mode

### PR DESCRIPTION
**Pull request by mustafa-gokce**
This PR adds pause/continue methods to mode auto and mode guided.
Pause/continue logic of the DO_FOLLOW and DO_REPOSITION commands in guided mode is implemented.
I will also implement other commands' pause/continue logic.

### CODE REVIEW

**Comment by peterbarker** [comment]
I asked @mustafa-gokce to create this so we can discuss it more easily.

I'm currently objecting to the current implementation as we can enter and use guided mode in ways *other* than receiving a mavlink command (e.g. via scripting).

There are two ways I can currently think of to do this better
 - have another SubMode of "pause" - but will we need to work out how to reinitialise old submode when we `continue`?
 - scribble down the values of every entry point into guided and have an enumeration for which one to call with we `continue`

@rmackay9 's input required.

**Comment by rmackay9** [comment]
I guess we should clarify what behaviour we want in Guided mode's various sub modes.

In the case where the pilot has triggered a takeoff or has provided a Location I think pausing and the restoring the target makes sense.  For the submodes where the vehicle is constantly being sent new targets from a Lua script or a companion computer we probably do **not** want to restore the old target.  Any submode with a timeout should not have the previous target restarted.

Another question is do we want to reject new commands passed in (from a Lua script or companion computer) if the vehicle has been "paused" in Guided mode?  I'm unsure if rejecting new commands would be useful or just annoying.

**Comment by mustafa-gokce** [comment]
> Any submode with a timeout should not have the previous target restarted.

@rmackay9 I thought it too but didn't sure about it. If a submode is without a timeout, should restart the behavior (should disable the mode for continuing or reissuing again). And also we should not resume the second latest submode command that has a timeout, I think.
> Another question is do we want to reject new commands passed in (from a Lua script or companion computer) if the vehicle has been "paused" in Guided mode? I'm unsure if rejecting new commands would be useful or just annoying.

@rmackay9 I think we should reset pause/continue when a new command is received (or we can say, started a new submode that is started by the method other than continuing) so the command will be accepted anyway.

I am open to all other suggestions, @rmackay9, @peterbarker.
Thanks for your time...

**Comment by mustafa-gokce** [comment]
With 51801b3, I removed support for pause/continue of the submodes that have no timeout.
Also without pausing first, any of the previous submodes will not be resumed.
Pausing a submode does not cause rejecting new commands.
In a mode or submode change, pause/continue will be restarted so resume can be only issued if submode is paused first.

File: ArduCopter/mode_auto.cpp [hunk 1]
```diff
@@ -2017,4 +2017,23 @@ bool ModeAuto::verify_nav_delay(const AP_Mission::Mission_Command& cmd)
     return false;
 }
 
+/********************************************************************************/
+// Pause and continue auto mode
+/********************************************************************************/
+
+bool ModeAuto::pause()
+{
+    mission.stop();
```
Comment (peterbarker) [inline ArduCopter/mode_auto.cpp:2026]: I wonder if we should do this *after* doing loiter_start....

File: ArduCopter/Copter.h [hunk 2]
```diff
@@ -626,7 +626,31 @@ class Copter : public AP_Vehicle {
...
+    // previous command types
+    enum class PrevCmd : uint8_t {
+        DO_USER_TAKEOFF_START = 0,
+        SET_DESTINATION_VECTOR = 1,
+        SET_DESTINATION_LOCATION = 2,
+        SET_DESTINATION_POSVEL = 3,
+        SET_DESTINATION_POSVELACCEL = 4,
+    };
 
+    // previous commands from GCS
```
Comment (peterbarker) [inline ArduCopter/Copter.h:638]: This state should all be contained within the ModeGuided class.

Same with prevCmd enum

File: ArduCopter/mode.cpp [hunk 3]
```diff
@@ -327,6 +327,9 @@ bool Copter::set_mode(Mode::Number mode, ModeReason reason)
         AP_Notify::events.user_mode_change = 1;
     }
 
+    // reset pause continue on mode changes
+    prev_cmd.paused = false;
```
Comment (peterbarker) [inline ArduCopter/mode.cpp:331]: Should go into Guided's "exit_mode" code.

File: ArduCopter/mode_guided.cpp [hunk 4]
```diff
@@ -106,6 +106,11 @@ bool ModeGuided::allows_arming(AP_Arming::Method method) const
 // do_user_takeoff_start - initialises waypoint controller to implement take-off
 bool ModeGuided::do_user_takeoff_start(float takeoff_alt_cm)
 {
+    // set previous command
+    copter.prev_cmd.prev_cmd_type = Copter::PrevCmd::DO_USER_TAKEOFF_START;
+    copter.prev_cmd.takeoff_alt_cm = takeoff_alt_cm;
+    copter.prev_cmd.paused = false;
```
Comment (peterbarker) [inline ArduCopter/mode_guided.cpp:112]:
```suggestion
    prev_cmd.prev_cmd_type = Copter::PrevCmd::DO_USER_TAKEOFF_START;
    prev_cmd.takeoff_alt_cm = takeoff_alt_cm;
    prev_cmd.paused = false;
```

similarly elsewhere

File: ArduCopter/mode_guided.cpp [hunk 5]
```diff
@@ -1177,4 +1238,61 @@ uint32_t ModeGuided::get_timeout_ms() const
...
+        }
+        case Copter::PrevCmd::SET_DESTINATION_POSVELACCEL: {
+            return set_destination_posvelaccel(copter.prev_cmd.destination, copter.prev_cmd.velocity,
+                                               copter.prev_cmd.acceleration, copter.prev_cmd.use_yaw,
+                                               copter.prev_cmd.yaw_cd, copter.prev_cmd.use_yaw_rate,
+                                               copter.prev_cmd.yaw_rate_cds, copter.prev_cmd.relative_yaw);
+        }
+        default: {
+            return false;
+        }
```
Comment (peterbarker) [inline ArduCopter/mode_guided.cpp:1293]:
```suggestion
```

Omitting this forces someone who adds a PrevCmd entry to deal with the continue case.  Generally when a value is constrained to being within an enumeration then we omit the `default` clause.

**Review by peterbarker — CHANGES_REQUESTED** [review]
Would be nice to get an autotest for this.

**Comment by mustafa-gokce** [comment]
> Would be nice to get an autotest for this.

Done.

File: ArduCopter/mode_guided.cpp [hunk 6]
```diff
@@ -37,6 +37,32 @@ struct Guided_Limit {
...
+// previous command types
+enum class PrevCmd : uint8_t {
+    DO_USER_TAKEOFF_START = 0,
+    SET_DESTINATION_VECTOR = 1,
+    SET_DESTINATION_LOCATION = 2,
+    SET_DESTINATION_POSVEL = 3,
+    SET_DESTINATION_POSVELACCEL = 4,
+};
+
+// previous commands from GCS
```
Comment (peterbarker) [inline ArduCopter/mode_guided.cpp:49]:
```suggestion
// current command information
```

**Comment by peterbarker** [comment]
Looking good.  Need @rmackay9 to weigh in again now.  He would probably like an outline of how you've tested this.

**Comment by mustafa-gokce** [comment]
To test this feature in GUIDED mode, I created an autotest.
What it does is:
1. Takeoff to an altitude in GUIDED mode.
2. Send move to a location command in GUIDED mode.
3. Wait some time to move away from home.
4. Send pause command.
5. Wait for ground speed to be zero.
6. Send continue command.
7. Wait vehicle to reach its destination.
8. Land and disarm.

File: ArduCopter/mode_guided.cpp [hunk 7]
```diff
@@ -1177,4 +1264,65 @@ uint32_t ModeGuided::get_timeout_ms() const
...
 }
 
+/********************************************************************************/
+// Pause and continue guided mode
+/********************************************************************************/
+
+// pause guide mode
+bool ModeGuided::pause()
+{
+    if(init(true)) {
```
Comment (peterbarker) [inline ArduCopter/mode_guided.cpp:1274]:
```suggestion
    if (init(true)) {
```

File: Tools/autotest/arducopter.py [hunk 8]
```diff
@@ -7431,6 +7431,73 @@ def PAUSE_CONTINUE(self):
 
         self.wait_disarmed()
 
+    # test pause/continue in guided mode
+    def PAUSE_CONTINUE_GUIDED(self):
+
+        destination = mavutil.location(-35.361286, 149.161107, 30, 0)
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py:7437]: This shouldn't really be hard-coded.  You can get a relative position with 
```
                loc = self.home_relative_loc_ne(n, e)
```

File: Tools/autotest/arducopter.py [hunk 9]
```diff
@@ -7431,6 +7431,73 @@ def PAUSE_CONTINUE(self):
...
+        self.run_cmd(mavutil.mavlink.MAV_CMD_DO_PAUSE_CONTINUE,
+                     1, # param1
+                     0, # param2
+                     0, # param3
+                     0, # param4
+                     0, # param5
+                     0, # param6
+                     0 # param7
+                     )
+
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py:7494]: I'd like to see more of the different destination types tested in here, and that changing out of guided / back into guided works appropriately.

**Review by peterbarker — CHANGES_REQUESTED** [review]
Please have a look at https://github.com/ArduPilot/ardupilot/commit/77fcef6ffedcb7ea32bf523827d79eb6ed707ced

It adds some degree of type-safety to this recording/replaying of parameters to the functions.  It's only half done; eventually no destination components should exist outside the per-destination-type structs.

File: ArduCopter/GCS_Mavlink.cpp [hunk 10]
```diff
@@ -991,30 +991,30 @@ MAV_RESULT GCS_MAVLINK_Copter::handle_command_long_packet(const mavlink_command_
 
 MAV_RESULT GCS_MAVLINK_Copter::handle_command_pause_continue(const mavlink_command_int_t &packet)
 {
-#if MODE_AUTO_ENABLED
-    if (copter.flightmode->mode_number() != Mode::Number::AUTO) {
-        // only supported in AUTO mode
+#if defined(MODE_AUTO_ENABLED) && defined(MODE_GUIDED_ENABLED)
```
Comment (mustafa-gokce) [inline ArduCopter/GCS_Mavlink.cpp:994]: I am not sure whether
`#if defined(MODE_GUIDED_ENABLED) && defined(MODE_AUTO_ENABLED)
`
or
`#if defined(MODE_GUIDED_ENABLED) || defined(MODE_AUTO_ENABLED)`

**Pull request closed without merging**